src/main/takeDetection/takeDetection.ts

```diff
@@ -48,56 +48,202 @@ export function findSentences(words: Word[]): Sentence[] {
-          takeGroups.at(-1)?.takes.push(nextTake);
-        } else {
-          takeGroups.push(newTakeGroup(sentences[i], sentences[i + 1]));
+
+  if (isSimilarTake)
+    takeGroups.push(
+      newTakeGroup(potentialTakeStartIdxs, potentialTakeLen, sentences)
+    );
+
+  return updatedCurrentSentenceIdx;
+};
+
+export function findTakes(words: Word[]): InjectableTakeGroup[] {
+  const sentences: Sentence[] = findSentences(words);
+  const takeGroups: InjectableTakeGroup[] = [];
+
+  let currentSentenceIdx = 0;
+  let detectNotCompelet = true;
+  const maxSentenceIdx = sentences.length - 1;
+
+  while (detectNotCompelet) {
+    let potentialTakeLen = 0;
+    const potentialTakeStartIdxs: number[] = [];
+
+    // set first sentence as potential take
+    potentialTakeStartIdxs.push(currentSentenceIdx);
+
+    let nextSentenceIdx = currentSentenceIdx + 1;
+    const maxRemainingTakesIdx =
+      Math.round((sentences.length - currentSentenceIdx) / 2) +
+      currentSentenceIdx;
+
+    // search for any similar sentence start from current first sentence
+    while (nextSentenceIdx < maxRemainingTakesIdx || potentialTakeLen > 0) {
+      // if remaining sentences can't make valid take
+      if (
+        nextSentenceIdx > maxSentenceIdx ||
+        nextSentenceIdx + potentialTakeLen - 1 > maxSentenceIdx
+      )
+        break;
+
+      const isSimilar =
+        getSimilarityScore(
+          sentences[currentSentenceIdx].sentenceString,
+          sentences[nextSentenceIdx].sentenceString
+        ) > THRESHOLD;
+
+      // if already found potentail take
+      // but next sentence at potential take index is not similar
+      if (potentialTakeLen > 0 && !isSimilar) break;
+
+      if (isSimilar) {
+        if (potentialTakeLen === 0) {
+          potentialTakeLen = nextSentenceIdx - currentSentenceIdx;
         }
+
+        potentialTakeStartIdxs.push(nextSentenceIdx);
+        nextSentenceIdx += potentialTakeLen;
       } else {
-        takeGroups.push(newTakeGroup(sentences[i], sentences[i + 1]));
+        nextSentenceIdx += 1;
       }
     }
+
+    if (potentialTakeStartIdxs.length > 1) {
+      if (potentialTakeLen > 1) {
+        // start detection after find all potential take start index
+        const updatedCurrentSentenceIdx = startDetection(
+          potentialTakeLen,
+          potentialTakeStartIdxs,
+          sentences,
+          takeGroups
+        );
+
+        currentSentenceIdx = updatedCurrentSentenceIdx;
+      } else {
+        takeGroups.push(
+          newTakeGroup(potentialTakeStartIdxs, potentialTakeLen, sentences)
+        );
+        currentSentenceIdx = potentialTakeLen * potentialTakeStartIdxs.length;
+      }
+    } else {
+      currentSentenceIdx += 1;
+    }
+
+    if (
+      potentialTakeStartIdxs[potentialTakeStartIdxs.length - 1] +
+        potentialTakeLen ===
+      sentences.length
+    ) {
+      detectNotCompelet = false;
```

Comment:
Since the only time you update this condition is at the end of the loop, it'd be cleaner as a `do ... while`